source/agora/consensus/EnrollmentManager.d

```diff
@@ -699,6 +734,100 @@ public class EnrollmentManager
         return true;
     }
 
+    /***************************************************************************
+
+        Check if revealing a pre-image is needed at a certain height
+
+        Params:
+            height = block height to check
+
+        Returns:
+            true if revealing a pre-image is needed
+
+    ***************************************************************************/
+
+    public bool needRevealPreimage (ulong height) @safe nothrow
+    {
+        if (height >= this.next_reveal_height)
+            return true;
+        else
+            return false;
+    }
+
+    /***************************************************************************
+
+        Increase the next reveal height by the revelation period
+
+    ***************************************************************************/
+
+    public void increaseNextRevealHeight () @safe nothrow
+    {
+        ulong next_height = this.getNextRevealHeight();
+        if (this.next_reveal_height < ulong.max)
```

Comment:
So if there is a database error then the node will silently stop emitting preimages? The only way the user will know this is if they're actively monitoring the logs. 

In the meantime the node stops emitting preimages, it gets slashed for not following the protocol, and the node owner loses funds. This is not good.

If there is an exceptional error that makes the node misbehave then we have to treat it as an error, not as a diagnostical issue. If the node can't reveal its preimages because of a database error then the node should shutdown with an error. The node operator can get notified when its node shuts down. That way the node operator gets the chance to fix the error (faulty disk or whatever else is causing the issue).

Reply:
> The node operator can get notified when its node shuts down.

Right. There needs a way to notify an abnormal status of a node to the operator through client UI.

Reply:
The UI is a separate concern. Agora doesn't have to run with any UI at all. It just has to exit with a process code `!= 0`, and then whichever initialization system is in use (systemd or otherwise) can notify the operator - whether it's through e-mail alerts, push notifications, etc. All of that stuff is external to the Agora node. But we must notify the operating system that we failed, and that means shutting down with an exit code.

Reply:
I think the operators of nodes need not to be familiar with command line environment. That's why I said about UI. Adding the code of assert or something in will be worked on one related PR or one issue.